allennlp/state_machines/beam_search.py

```diff
@@ -33,15 +36,38 @@ class BeamSearch(FromParams, Generic[StateType]):
         more diversity into the search. See Freitag and Al-Onaizan 2017,
         "Beam Search Strategies for Neural Machine Translation".
     """
-    def __init__(self, beam_size: int, per_node_beam_size: int = None) -> None:
+    def __init__(self,
+                 beam_size: int,
+                 per_node_beam_size: int = None,
+                 initial_sequence: torch.Tensor = None,
+                 keep_beam_details: bool = False) -> None:
         self._beam_size = beam_size
         self._per_node_beam_size = per_node_beam_size or beam_size
 
+        if initial_sequence is not None:
+            # construct_prefix_tree wants a tensor of shape (batch_size, num_sequences, sequence_length)
+            # so we need to add the first two dimensions in. This returns a list, but we're assuming
+            # batch size 1, so we extract the first element.
+            self._allowed_transitions = util.construct_prefix_tree(initial_sequence.view(1, 1, -1))[0]
+        else:
+            self._allowed_transitions = None
+
+        if keep_beam_details:
+            self.beams: List[List[Tuple[float, List[int]]]] = []
```

Comment:
I was pretty confused about what's going on here, because you're mixing batch index and number of steps below, and it seemed really hard to reason about.  Then I realized that you're assuming you only have batch size of 1 when you keep this around.  I'd probably make this a dictionary going from batch index to this list.  Monstrous type, but at least it doesn't break if you have batched computation.  Or have a major caveat somewhere that this is only intended for use with batch size 1.  This should go in the docstring for `keep_beam_details`.  Related: that parameter and `initial_sequence` need to be added to the class docstring.

Also, regarding terminology: you're right that this keeping snapshots of the beam at every timestep, so something like "beams" isn't as funny sounding as it otherwise might be.  But maybe `beam_snapshots` or something similar would be more obvious?